annotated.go

```diff
@@ -89,3 +91,296 @@ func (a Annotated) String() string {
+		}
+		annotatedParams = append(annotatedParams, structField)
+	}
+
+	ann.Ins = []reflect.Type{reflect.StructOf(annotatedParams)}
+	return nil
+}
+
+// ParamTags is an Annotation that annotates the parameter(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional parameter.
+func ParamTags(tags ...string) Annotation {
+	return paramTagsAnnotation{tags}
+}
+
+type resultTagsAnnotation struct {
+	tags []string
+}
+
+var _ resultTagsAnnotation = resultTagsAnnotation{}
+
+// Given func(T1, T2, T3, ..., TN), this generates a type roughly
+// equivalent to,
+//
+//   struct {
+//     fx.Out
+//
+//     Field1 T1 `$tags[0]`
+//     Field2 T2 `$tags[1]`
+//     ...
+//     FieldN TN `$tags[N-1]`
+//   }
+func (rt resultTagsAnnotation) apply(ann *annotations) error {
+	if ann.annotatedOut {
+		return fmt.Errorf("cannot apply more than one line of ResultTags")
+	}
+	ann.annotatedOut = true
+	fType := ann.fType
+
+	annotatedResult := []reflect.StructField{{
+		Name:      "Out",
+		Type:      reflect.TypeOf(Out{}),
+		Anonymous: true,
+	}}
+
+	returnsError := false
+	for i := 0; i < fType.NumOut(); i++ {
+		if fType.Out(i) == _typeOfError {
+			returnsError = true
+			continue
+		}
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: fType.Out(i),
+		}
+		if i < len(rt.tags) {
+			structField.Tag = reflect.StructTag(rt.tags[i])
+		}
+		annotatedResult = append(annotatedResult, structField)
+	}
+	if returnsError {
+		ann.Outs = []reflect.Type{
+			reflect.StructOf(annotatedResult),
+			_typeOfError,
+		}
+	} else {
+		ann.Outs = []reflect.Type{reflect.StructOf(annotatedResult)}
+	}
+	return nil
+}
+
+// ResultTags is an Annotation that annotates the result(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional result.
+func ResultTags(tags ...string) Annotation {
+	return resultTagsAnnotation{tags}
+}
+
+// Annotate lets you annotate a function's paramter and returns with tags
+// without you having to declare separate struct definitions for them.
+//
+// For example,
+//   func NewGateway(ro, rw *db.Conn) *Gateway { ... }
+//   fx.Provide(
+//     fx.Annotate(
+//       NewGateway,
+//       fx.ParamTags(`name:"ro" optional:"true"`, `name:"rw"`),
+//       fx.ResultTags(`name:"foo"`),
+//     ),
+//   )
+//
+// Is equivalent to,
+//
+//  type params struct {
+//    fx.In
+//
+//    RO *db.Conn `name:"ro" optional:"true"`
+//    RW *db.Conn `name:"rw"`
+//  }
+//
+//  type result struct {
+//    fx.Out
+//
+//    GW *Gateway `name:"foo"`
+//   }
+//
+//   fx.Provide(func(p params) result {
+//     return result{GW: NewGateway(p.RO, p.RW)}
+//   })
+//
+// If a single annotation appears multiple times, all annotations
+// will be ignored.
+//
+// For example,
+//
+//  fx.Provide(
+//    fx.Annotate(
+//      NewGateWay,
+//      fx.ParamTags(`name:"ro" optional:"true"`),
+//      fx.ParamTags(`name:"rw"),
```

Comment:
Worth highlighting that this is bad:
```suggestion
//      fx.ParamTags(`name:"rw"),  // ERROR: ParamTags was already used above
```